README.md
# sentinel2

[![Build Status](https://travis-ci.org/IVFL-BOKU/sentinel2.svg?branch=master)](https://travis-ci.org/IVFL-BOKU/sentinel2)
[![Coverage Status](https://coveralls.io/repos/github/IVFL-BOKU/sentinel2/badge.svg?branch=master)](https://coveralls.io/github/IVFL-BOKU/sentinel2?branch=master)

Accessing https://s2.boku.eodc.eu Sentinel-2 API from R made easy.

## Getting Started

This document aims to provide basic guidance on the use of the `sentinel2` package for R.
The following sections will cover the following steps:

   1. **Installation**
   2. **User login** to get **access** to the data portal
   3. **Query the database** for available **Sentinel-2 data products**
   4. **Request processing** of Sentinel-2 data products for a **region of interest**
   5. **Download** data products


## Installation

The `devtools` package allows to easily install `sentinel2` directly from GitHub, so make sure you have 
`devtools` installed.

```r
# If you don't have it yet, install 'devtools':
# install.packages('devtools')
devtools::install_github('IVFL-BOKU/sentinel2')
```

## User Login

In order to **get access** to the data portal, you have to login as a user. If you 
are not yet a **registered user**, please visit [s2.boku.eodc.eu](https://s2.boku.eodc.eu/) 
first.

Interaction with the data portal typically requires passing user credentials.
Use `S2_initialize_user()` to store your user credentials for the active session, 
and subsequently use `S2_check_access()` to verify access to the data portal.
The first lines of your code make look like this:

```r
library(sentinel2)

S2_initialize_user(user = 'test@s2.boku.eodc.eu', password = 'test')
S2_check_access()
```

Now we have access to the data portal! We might want to query for available data 
now.

## Query the database for Sentinel-2 data products

The data portal ([s2.boku.eodc.eu](https://s2.boku.eodc.eu/)) allows to obtain 
atmospherically corrected Sentinel-2 data and value added products 
(e.g. *LAI*, *AOT*, *FAPAR*, ...) on request. The processed data can be obtain 
in several ways, e.g. as *granules* or *single band images*.

### Query for *granules*

Suppose you are interested in atmospherically corrected data available at a 
specific location (e.g. *LONG 48.15, LAT 16.20*) for a given period of time 
(e.g. *'2016-09-01' to '2016-09-30'*). We will use `S2_query_granule()` to see, 
if there are already some processed data available:

```r
S2_query_granule(atmCorr  = TRUE, 
                 geometry = c(x = 16.20, y = 48.15), 
                 dateMin  = '2016-09-15', 
                 dateMax  = '2016-09-30')
```

The `data.frame` returned contains information about the granules found in the 
database, matching the requested criteria. There are several atmospherically 
corrected granules available, however, none are owned by the user, as the `NA`'s 
in the `url` column reveal.

*Coins* are used to gain access to individual granules using its *regionId* in 
`S2_buy_granule()`:

```r
S2_buy_granule(granuleId = 1080943)
```

*(Please take a look at [s2.boku.eodc.eu](https://s2.boku.eodc.eu/wiki/#!granule.md),
if you are unclear about the meaning of the columns and for a complete list of 
available query options).*

### Query for individual *images*

If we are interested in *indivdual bands* of Sentinel-2 or readily processed 
products such as *LAI*, we can use `S2_query_image()`:

```r
S2_query_image(band        = 'LAI', 
               utm         = '33U', 
               dateMin     = '2016-09-15', 
               dateMax     = '2016-09-30')
```

Although we found processed data in this case (, because another user has 
previously requested processing of the desired area), we expect the returned 
`data.frame` to be empty. Data processing is done on request, so we will need to 
place an order first.

## **Request processing** of Sentinel-2 data products for a **region of interest**

Data processing can be requested via `S2_put_roi()` for a specific location 
(provided e.g. as LAT/LONG point coordinates, a JSON geometry string or objects 
of class `SpatialPoints` or `SpatialPolygons`) and a given period of time. 

To issue a processing request it might be more convenient for you use the web 
interface at [s2.boku.eodc.eu](https://s2.boku.eodc.eu/). Be aware that this 
step requires some consciousness regarding the amount of data that is going to 
be processed!

Placing an order via `S2_put_ROI()` could look like this:

```r
S2_put_ROI(geometry    = c(x=16, y=48),
           regionId    = 'testROI', 
           cloudCovMax = 20, 
           dateMin     = '2016-06-01', 
           dateMax     = '2016-07-01')
```

Since data processing over large areas is somewhat costly, you need to pay 
coins if you issue processing for a region of interest.

`S2_user_info()` allows to check the budget of the user currently logged in:

```r
S2_user_info()
```

In this case, we used the 'test@s2.boku.eodc.eu' user. As we can see, all the 
coins have already been used!

## Download data

Downloading data from the portal can be done using `S2_download()`. You simply 
to pass the `url` found in the `data.frame` returned by `S2_query_granule()` or 
`S2_query_image()` to `S2_download()` and specify filenames / filepaths to save 
the data to.

### Downloading granules

Suppose you want to download granules you already own. We can find all granules 
owned by a user via `S2_query_granule(owned = TRUE)`:
```r
granules_owned <- S2_query_granule(owned = TRUE)
granules_owned

```

The column 'granules_owned$url' contains the path to the data we are aboeut to 
download. For each element we download we will have to specify a filename. You 
can either use your own naming scheme or use the convinience function 
`S2_generate_names()` to automatically create names from the 'data.frame' 
returned by the query:

```r
save_names <- S2_generate_names(x = granules_owned)
save_names
```

Now we can simply download the data (in this case to the current working 
directory) like so:

```r
S2_download(url = granules_owned$url, destfile = save_names)
```


### Downloading images

Downloading individual images can be done similarly. First we must find some 
images we own. Let's assume we are only interested in Band 8 and restrict the 
cloud coverage to be at most 85 percent:

```r
images = S2_query_image(owned = TRUE, band = 'B08', cloudCovMax = 85)
```

Then we need to create file names for them.
Lets do that by simply pasting acquisition date and file format.

```r
save_names = paste0(images$date, '.', images$format)
```

Finally we can download the images:

```r
S2_download(images$url, save_names)
```

#### Advanced image download

The image API allows you to perform many transformations on the downloded image, e.g.:

* reproject the image
* change the resolution
* change the data format (e.g. to float)
* adjust values range (e.g. divide by 10 so they will fit in the Byte values range)
* crop image to a given geometry

All options are described in the (API doc)[https://s2.boku.eodc.eu/wiki/#!image.md#GET_https://s2.boku.eodc.eu/image/{imageId}].

To apply a transformation you need to pass a corresponding API parameter to the `S2_download()` function call - see example below.

```r
# find some images and prepare file names
images = S2_query_image(owned = TRUE, band = 'B08', cloudCovMax = 85)
file_names = paste0(images$date, '.tif')

# read the geometry from file
geom = roi_to_jgeom('/my/path/my_geom_file.kml')

# download them:
# - reporojecting to WGS-84 (srid 4326)
# - changing data format to Byte (0-255)
# - dividing all values by 20 so they will better fit the Byte range 
#   (and setting max value to 254 so it will not overlap with the no data value)
# - cutting to the given geometry
S2_download(images$url, file_names, srid = 4326, dataType = 'Byte', range = 50, max = 254, geometry = geom)
```


NEWS.md
# 0.5.4 (2022-05-10)

* supress display of the "CRS object has comment, which is lost in output" warning 
  generated by the `sp::proj4string()` in the `roi_to_jgeom()`

# 0.5.3 (2020-08-03)

* fixed dplyr 1.0.0 deprecation notes

# 0.5.2 (2019-05-06)

* make `S2_initialize_user()` work with `user` and/or `password` being named vectors

# 0.5.1 (2019-04-29)

* `S2_download()` fixes:
    * `TRUE` is reported for local files skipped on `skipExisting = 'always'` condition
    * zip extraction is performed only if the response content type suggests it's a zip

# 0.5.0 (2019-04-29)

* `S2_download()` enhancements:
    * `tries` parameter added allowing automatic error handling
    * `skipExisting` parameter is tri-state now:
        * `always` downloads the data only if local copy doesn't exist 
           (no matter its size)
        * `samesize`  downlads the data if local copy doesn't exist or has different size
          (be aware it doesn't work for full granule zip downloads, in such a case it's
           equivalent to `never`)
        * `never` downloads the data no matter if local copy exists
    * `timeout` now defaults to 1800 (seconds) which seems to be big-enough and prevents
      download jobs from getting stalled

# 0.4.3 (2019-04-26)

* Fixes in `S2_download()`:
    * progress bar is now updated after each file download
    * progress bar is now properly updated no matter if a file exists locally or not
    * a `timeout` parameter can be specified limiting single file download time

# 0.4.2 (2019-03-18)

* A minor error generating warnings in `S2_buy_granule()` fixed.

# 0.4.1 (2019-01-29)

* Make `geojson_to_geometry()` public.

# 0.4.0 (2019-01-04)

Backward incompatible changes:

* `S2_put_ROI()` renamed to `S2_put_roi()` to follow naming convention of other 
  functions.
* `S2_put_roi()` nor returns parsed REST API response object instead of the raw
  `httr` package response object.
* Default `dateMin` and `dateMax` values removed from `S2_query_roi()`.

Enhancements:

* `S2_query_roi()` supports the `spatial` parameter allowing to parse ROIs'
  geometries into R spatial objects.
* `S2_do_query()` casts returned data frames to `dplyr's` tibble if the `dplyr` 
  package is installed.
* `S2_do_query()` provides information useful for debugging if a request fails.
* `roi_to_jgeom()` made more flexible (can handle all `sp's` package `*DataFrame` 
  objects as well as reproject objects in non WGS-84 projections)
* Tests coverage rised slightly.
* Credentials are now stored in a little safer way (in a package's private environment
  instead of global R options).

Bugfixes:

* Fixes to `S2_buy_granules()`.

Other:

* Obsolete internal functions removed

# 0.3.1 (2018-12-18)

* It is now possible to abort the `S2_download()` function.
* A `spatial` parameter of the `S2_query_granule()` and `S2_query_image()` is now
  a string and allows to choose between the `sp` package and the `sf` package spatial
  object representation.
* All `S2_query_...()` functions always return a data frame even when query fetches no
  results. Returned data frame is guaranteed to have at least an `{objectType}Id` 
  column (e.g. `granuleId` for granules, `regionId` for regions of interest, etc.).
* The `S2_download()` can now display a progress bar.

# 0.2.1 (2018-10-25)

* A `regionId` parameter added to all `S2_query_...()` functions for which it's valid.
  While since 0.1.0 it was possible to use it using the `...` parameter, the explicit
  form allows to document it nicely.
* Default values for `dateMin`, `dateMax`, `cloudCovMin` and `cloudCovMax` parameters
  dropped from all `S2_query_...()` functions. They were useless (simply followed 
  default values assumed by the underlaing REST API) and they messed up proper `regionId`
  parameter handling (they interfered with picking up default `dateMin`, `dateMax` and
  `cloudCovMax` values from the region of interest settings)
* Tests coverage slightly improved

# 0.1.1 (2018-06-21)

* `S2_put_ROI()` now also works for admin users.

# 0.1.0 (2017-07-31)

* `S2_download()` was extended with `...` allowing to pass any additional
  parameters to the HTTP API.  
  See the *Advanced image download* chapter in the *introduction* vignette.
* `S2_query_granule()` always returns a data.frame.
  The `spatial` parameter now affects only the type of the `geometry` column
  in the returned data.frame (GeoJSON strings when `FALSE`, 
  `SpatialPolygonsDataFrame` objects when `TRUE`).
* `S2_query_image()` accepts the `spatial` parameter which is triggering conversion
  of returned GeoJSON geometry strings into `SpatialPolygonsDataFrame` objects.
* `spat_to_jgeom()` is now exported and uses `rgdal::readOGR()` in the back.  
  This allows to simply convert any vector file into a GeoJSON geometry string
  accepted by the HTTP API calls.


R/S2_put_roi.R
#' Creates new or update existing regions of interest
#'
#' Place/Update roi for processing Sentinel-2 data
#'
#' @param geometry SpatialPoints, SpatialPolygons or path to a shapefile on disk
#' @param regionId character a new or existing region name
#' @param cloudCovMax integer cloud coverage treshold for a given region
#' @param dateMin character, date 'YYYY-MM-DD' beginning of the region's time span
#' @param dateMax character, date 'YYYY-MM-DD' end of the region's time span
#' @param indicators character vector of indicator names to be computed for a
#'   given region (e.g. \code{c("LAI", "FAPAR")})
#' @param srid integer geometry projection SRID (e.g. 4326 for WGS-84)
#' @return object describing the created ROI (as returned by the REST API)
#' @export
S2_put_roi = function(
  geometry,
  regionId    = NULL,
  cloudCovMax = 50,
  indicators  = NULL,
  dateMin     = NULL,
  dateMax     = NULL,
  srid        = 4326
) {
  if (is.null(dateMin) || is.null(dateMax)) {
    stop("Please supply 'dateMin' and 'dateMax' in format 'YYYY-MM-DD")
  }

  if (check_date(dateMin) > check_date(dateMax)) {
    stop("'dateMin' (", dateMin, ") larger than 'dateMax' (", dateMax, ")")
  }

  if (is.null(regionId)) {
    stop(
      "'regionId' not specified!",
      "\n-> If you want to update an existing 'roi', please supply valid 'regionId'",
      "\n-> If you like to create a new 'regionId' enter desired name"
    )
  }

  geometry  = roi_to_jgeom(geometry)
  body_l    = list(
    cloudCovMax  = cloudCovMax,
    dateMin      = dateMin,
    dateMax      = dateMax,
    geometry     = geometry,
    indicators   = indicators,
    srid         = srid
  )
  body_l   = body_l[!sapply(body_l , is.null)]

  credentials = get_credentials()
  url = httr::modify_url(
    'https://s2.boku.eodc.eu',
    path   = list("user", credentials['user'], "roi", regionId)
  )
  rtrn = httr::PUT(
    url = url, body = body_l,
    encode = 'json',
    config = httr::authenticate(credentials['user'], credentials['password'])
  )

  return(jsonlite::fromJSON(httr::content(rtrn, as = 'text')))
}
